services/src/main/java/org/keycloak/services/resources/account/AccountCredentialResource.java

```diff
@@ -113,104 +126,103 @@ public boolean isRemoveable() {
-        dummyCreds.add(passwordless);
-        
-        return dummyCreds;
+
+        boolean filterUserCredentials = userCredentials != null && !userCredentials;
+
+        List<CredentialContainer> credentialTypes = new LinkedList<>();
+        List<CredentialProvider> credentialProviders = UserCredentialStoreManager.getCredentialProviders(session, realm, CredentialProvider.class);
+        Set<String> enabledCredentialTypes = getEnabledCredentialTypes(credentialProviders);
+
+        List<CredentialModel> models = filterUserCredentials ? null : session.userCredentialManager().getStoredCredentials(realm, user);
+
+        // Don't return secrets from REST endpoint
+        if (models != null) {
+            for (CredentialModel credential : models) {
+                credential.setSecretData(null);
+            }
+        }
+
+        for (CredentialProvider credentialProvider : credentialProviders) {
+            String credentialProviderType = credentialProvider.getType();
+
+            // Filter just by single type
+            if (type != null && !type.equals(credentialProviderType)) {
+                continue;
+            }
+
+            boolean enabled = enabledCredentialTypes.contains(credentialProviderType);
+
+            // Filter disabled credential types
+            if (!enabled) {
+                continue;
+            }
+
+            CredentialTypeMetadata metadata = credentialProvider.getCredentialTypeMetadata();
+
+            List<CredentialRepresentation> userCredentialModels = filterUserCredentials ? null : models.stream()
+                    .filter(credentialModel -> credentialProvider.getType().equals(credentialModel.getType()))
+                    .map(ModelToRepresentation::toRepresentation)
+                    .collect(Collectors.toList());
+
+            CredentialContainer credType = new CredentialContainer(metadata, userCredentialModels);
+            credentialTypes.add(credType);
+        }
+
+        credentialTypes.sort(Comparator.comparing(CredentialContainer::getMetadata));
+
+        return credentialTypes;
     }
-//
-//
-//    @GET
-//    @Path("registrators")
-//    @NoCache
-//    @Produces(javax.ws.rs.core.MediaType.APPLICATION_JSON)
-//    public List<String> getCredentialRegistrators(){
-//        auth.requireOneOf(AccountRoles.MANAGE_ACCOUNT, AccountRoles.VIEW_PROFILE);
-//
-//        return session.getContext().getRealm().getRequiredActionProviders().stream()
-//                .map(RequiredActionProviderModel::getProviderId)
-//                .filter(providerId ->  session.getProvider(RequiredActionProvider.class, providerId) instanceof CredentialRegistrator)
-//                .collect(Collectors.toList());
-//    }
-//
+
+    // Going through all authentication flows and their authentication executions to see if there is any authenticator of the corresponding
+    // credential type.
+    private Set<String> getEnabledCredentialTypes(List<CredentialProvider> credentialProviders) {
```

Comment:
This is true and I was also thinking about this. Problem is, that there can be various corner cases. For example:
- Flow with WebAuthn authenticator could be configured as "First broker login" flow
- Flow with WebAuthn authenticator could be configured as "Authetnication flow binding override" for any client (and some of our users has 10K clients in the realm).

 I suppose that when admin configured some flow with WebAuthn authenticator, he actually wants to use WebAuthn somewhere. So I've just used this algorithm as some compromise between accuracy and performance (EG avoid to loop through all the clients and their binding overrides etc). I hope it works as good compromise if we properly document it?

Reply:
@mposolda Thanks for the explanation. That's right and now makes a complete sense.

Reply:
Thinking about this a bit more I believe it might be a problem for OTP. Yes, WebAuthn this approach works but what if admin doesn't want to enable OTP? Users will always see the OTP option in the account console because the OTP authenticator is always present in the built-in browser flow which can't be deleted. Or am I missing something?

Reply:
Yes. I think in case that OTP authenticator is disabled or it is used in the disabled subflow, it won't be used. That should work fine for built-in browser flow as well, but we need to document it. WDYT?

It will be ideal if there is some way for disable some credential types entirely in some simple way - maybe just switch in admin console or something like that, which will simply allow to enable/disable "otp" or "webauthn" credential etc. But that won't be for 9.0.0...

Reply:
> Yes. I think in case that OTP authenticator is disabled or it is used in the disabled subflow, it won't be used. That should work fine for built-in browser flow as well, but we need to document it. WDYT? I will need to update PR for this as it currently doesn't take "disabled" into account.

+1
> It will be ideal if there is some way for disable some credential types entirely in some simple way - maybe just switch in admin console or something like that, which will simply allow to enable/disable "otp" or "webauthn" credential etc. But that won't be for 9.0.0...

+1000 :D
However, we might want to wait with that for the new Admin Console where auth flows should be redesigned anyway.

Reply:
I've updated the PR to behave like that and don't consider disabled authenticators. Added test that it is possible to disable OTP and not return it from account REST credential endpoint